internal/xds/translator/listener.go

```diff
@@ -149,9 +189,10 @@ func (t *Translator) addXdsHTTPFilterChain(xdsListener *listenerv3.Listener, irL
 	}
 
 	// Client IP detection
-	var xffNumTrustedHops uint32
-	if irListener.ClientIPDetection != nil && irListener.ClientIPDetection.XForwardedFor != nil {
-		xffNumTrustedHops = ptr.Deref(irListener.ClientIPDetection.XForwardedFor.NumTrustedHops, 0)
+	var useRemoteAddress = true
+	var originalIPDetectionExtensions = originalIPDetectionExtensions(irListener.ClientIPDetection)
+	if originalIPDetectionExtensions != nil {
+		useRemoteAddress = false
```

Comment:
does this have to be false ?
this knobs helps the next upstream with middleware client ip info, unsure how its tied to downstream client ip info which is being used to determine true client ip for logging, acl, ratelimiting

Reply:
> does this have to be false? this knobs helps the next upstream with middleware client ip info

It has to be `false`. Both top-level `xff_num_trusted_hops` and `use_remote_address` are mutually exclusive with `original_ip_detection_extensions`. There's a note in the docs about it, and if you try setting it to `true` while also configuring `original_ip_detection_extensions` envoy will reject the config:

```
[2024-02-07 22:30:29.118][1][warning][config] [source/extensions/config_subscription/grpc/grpc_subscription_impl.cc:138] gRPC config for type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) default/eg/http: Original IP detection extensions and use_remote_address may not be mixed   
```

And yes, this does mean when `customHeader` is configured the `X-Forwarded-For` header will not have the envoy proxy IP appended to it since `use_remote_address` is `false` and it's assumed you aren't relying on the XFF headers when using alternate forms of client IP detection.

> unsure how its tied to downstream client ip info which is being used to determine true client ip for logging, acl, ratelimiting

The value of the named header will be present in log entries as the `downstream_remote_address`, so I would assume that it would still be used for ACLs and rate limits.

`AllowExtensionToSetAddressAsTrusted: true` is set on the extension which at minimum allows the address to also be used to determine if the address is "internal" or not based on the CIDR range.

Reply:
thanks for explaining this